platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/DataFileWriter.java

```diff
@@ -211,20 +162,127 @@ public synchronized long storeDataItem(final Consumer<BufferedData> dataItemWrit
+            this.fileStartOffset = fileStartOffset;
+
+            // ensure we have disk space buffer can use
+            file.setLength(fileStartOffset + bufferSize);
+
+            final MappedByteBuffer newMapping = file.getChannel().map(MapMode.READ_WRITE, fileStartOffset, bufferSize);
+            if (newMapping == null) {
+                throw new IOException("Failed to map file channel to memory");
+            }
+
+            close();
+
+            bufferFileMapping = newMapping;
+            bufferedData = BufferedData.wrap(bufferFileMapping);
+        }
+
+        long getCurrentFilePosition() {
+            return fileStartOffset + bufferedData.position();
+        }
+
+        /**
+         * @param field field describing writing data
+         * @param expectedSize expected data size that will be checked against the actual size after writing
+         * @param writer writer of the data
+         * @return the file offset where the data was written (start position of the data item)
+         * @throws IOException
+         */
+        long write(FieldDefinition field, int expectedSize, Consumer<BufferedData> writer) throws IOException {
+            long curPos = bufferedData.position();
+            final long fileOffset = getCurrentFilePosition();
+            final int sizeToWrite = ProtoWriterTools.sizeOfDelimited(field, expectedSize);
+
+            if (sizeToWrite > bufferSize) {
+                throw new IOException(DataFileCommon.ERROR_DATAITEM_TOO_LARGE + " dataSize=" + sizeToWrite
+                        + ", bufferSize=" + bufferSize);
+            }
+
+            // if there is not enough space in the current mapped buffer, we need to move it to start at current file
+            // offset
+            if (bufferedData.remaining() < sizeToWrite) {
+                move(fileOffset);
+                curPos = 0;
+            }
+
+            try {
+                ProtoWriterTools.writeDelimited(bufferedData, field, expectedSize, writer);
+                if (bufferedData.position() != curPos + sizeToWrite) {
+                    throw new IOException("Estimated size / written bytes mismatch: expected=" + sizeToWrite
+                            + " written=" + (bufferedData.position() - curPos));
+                }
+            } catch (final BufferOverflowException e) {
+                // Buffer overflow here means the mapped buffer is smaller than even a single data item
+                throw new IOException(DataFileCommon.ERROR_DATAITEM_TOO_LARGE, e);
+            }
+            return fileOffset;
+        }
+
+        @Override
+        public void close() {
+            if (bufferFileMapping != null) {
+                // TODO should we flush changes to the disk here for durability that takes little additional time
+                // or we rely on async flush in the OS?
+                // bufferFileMapping.force();
+                MemoryUtils.closeMmapBuffer(bufferFileMapping);
```

Comment:
Set `bufferFileMapping` (and `bufferedData`) to null to protect from double closing